File: frontend/src/lib/training.ts
/** Training stages + modules admin API client. */

import { apiFetch } from "@/lib/api";
import type { AiLevel } from "@/lib/knowledge";

type Json = Record<string, unknown>;

export interface Stage {
  id: string;
  key: string;
  name: string;
  description: string | null;
  order_index: number;
  entry_conditions: Json | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface StageInput {
  key?: string;
  name?: string;
  description?: string | null;
  order_index?: number;
  entry_conditions?: Json | null;
  is_active?: boolean;
}

export interface Module {
  id: string;
  key: string;
  name: string;
  target_user: string | null;
  stage_id: string | null;
  goal: string | null;
  steps: string[];
  duration_minutes: number | null;
  stop_conditions: Json | null;
  next_module_id: string | null;
  next_stage_id: string | null;
  min_ai_level: AiLevel;
  order_index: number;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface ModuleInput {
  key?: string;
  name?: string;
  target_user?: string | null;
  stage_id?: string | null;
  goal?: string | null;
  steps?: string[];
  duration_minutes?: number | null;
  stop_conditions?: Json | null;
  next_module_id?: string | null;
  next_stage_id?: string | null;
  min_ai_level?: AiLevel;
  order_index?: number;
  is_active?: boolean;
}

const STAGES = "/api/admin/stages";
const MODULES = "/api/admin/training-modules";

export const listStages = () => apiFetch<Stage[]>(STAGES, { auth: true });
export const createStage = (input: StageInput) =>
  apiFetch<Stage>(STAGES, { method: "POST", body: input, auth: true });
export const updateStage = (id: string, input: StageInput) =>
  apiFetch<Stage>(`${STAGES}/${id}`, { method: "PATCH", body: input, auth: true });
export const deleteStage = (id: string) =>
  apiFetch<void>(`${STAGES}/${id}`, { method: "DELETE", auth: true });

export const listModules = () => apiFetch<Module[]>(MODULES, { auth: true });
export const createModule = (input: ModuleInput) =>
  apiFetch<Module>(MODULES, { method: "POST", body: input, auth: true });
export const updateModule = (id: string, input: ModuleInput) =>
  apiFetch<Module>(`${MODULES}/${id}`, { method: "PATCH", body: input, auth: true });
export const deleteModule = (id: string) =>
  apiFetch<void>(`${MODULES}/${id}`, { method: "DELETE", auth: true });

File: backend/app/schemas/training.py
"""Training stage and module request/response schemas."""

from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import AiLevel

# --------------------------------- stages ---------------------------------


class StageCreate(BaseModel):
    key: str = Field(min_length=1, max_length=60)
    name: str = Field(min_length=1, max_length=120)
    description: str | None = None
    order_index: int = 0
    entry_conditions: dict[str, Any] | None = None
    is_active: bool = True


class StageUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=120)
    description: str | None = None
    order_index: int | None = None
    entry_conditions: dict[str, Any] | None = None
    is_active: bool | None = None


class StageRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    key: str
    name: str
    description: str | None
    order_index: int
    entry_conditions: dict[str, Any] | None
    is_active: bool
    created_at: datetime
    updated_at: datetime


# -------------------------------- modules ---------------------------------


class ModuleCreate(BaseModel):
    key: str = Field(min_length=1, max_length=60)
    name: str = Field(min_length=1, max_length=120)
    target_user: str | None = Field(default=None, max_length=160)
    stage_id: UUID | None = None
    goal: str | None = None
    steps: list[str] = Field(default_factory=list)
    duration_minutes: int | None = Field(default=None, ge=0)
    stop_conditions: dict[str, Any] | None = None
    next_module_id: UUID | None = None
    next_stage_id: UUID | None = None
    min_ai_level: AiLevel = AiLevel.none
    order_index: int = 0
    is_active: bool = True


class ModuleUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=120)
    target_user: str | None = Field(default=None, max_length=160)
    stage_id: UUID | None = None
    goal: str | None = None
    steps: list[str] | None = None
    duration_minutes: int | None = Field(default=None, ge=0)
    stop_conditions: dict[str, Any] | None = None
    next_module_id: UUID | None = None
    next_stage_id: UUID | None = None
    min_ai_level: AiLevel | None = None
    order_index: int | None = None
    is_active: bool | None = None


class ModuleRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    key: str
    name: str
    target_user: str | None
    stage_id: UUID | None
    goal: str | None
    steps: list[str]
    duration_minutes: int | None
    stop_conditions: dict[str, Any] | None
    next_module_id: UUID | None
    next_stage_id: UUID | None
    min_ai_level: AiLevel
    order_index: int
    is_active: bool
    created_at: datetime
    updated_at: datetime
